File: Add Code Here/PYTHON/Trapping_Rain_Water.py
"""We are given with n non-negative integers representing an elevation map where the width of each bar is 1,
we need to compute how much water it is able to trap after raining."""

def rain(arr):
    ans = 0
    temp = 0
    prev = 0
    for i in range(len(arr)):
        if arr[i] > prev and temp == 0:
            prev = arr[i]
        elif arr[i] >= prev:
            ans += temp
            prev = arr[i]
            temp = 0
        else:
            if i != len(arr)-1 and arr[i] < max(arr[i+1:]):
                temp += prev - arr[i]
            else:
                ans += arr[i]
                temp = 0
                prev = arr[i]
    return ans


arr = [2, 0, 3, 0, 2, 0, 4]
print(rain(arr))

File: Add Code Here/Java/Trapping_Rain_Water.java
//leetcode - Trapping Rain Water
public class Trapping_Rain_Water {
    public static void main(String[] args) {
        int[] height = {0,1,0,2,1,0,1,3,2,1,2,1};
        System.out.println(trap(height));
    }
    public static int trap(int[] height) {
        int n = height.length;
        int [] left = new int[n];
        int [] right = new int[n];
        left[0] = height[0];
        for(int i = 1;i<n;i++){
            left[i] = Math.max(left[i-1],height[i]);
        }
        right[n-1] = height[n-1];
        for(int i = n-2;i>=0;i--){
            right[i] = Math.max(right[i+1],height[i]);
        }
        int ans = 0;
        for(int i =0;i<n;i++){
            ans = ans + (Math.min(left[i],right[i])-height[i]);
        }
        return ans;
    }
}
